### ISSUE TITLE
HTTP 403 rate limit exceeded when updating (-U)

### ISSUE THREAD

**Issue by cheater** [yt-dlp/yt-dlp#8175]
### DO NOT REMOVE OR SKIP THE ISSUE TEMPLATE

- [X] I understand that I will be **blocked** if I *intentionally* remove or skip any mandatory\* field

### Checklist

- [X] I'm reporting a bug unrelated to a specific site
- [X] I've verified that I'm running yt-dlp version **2023.07.06** ([update instructions](https://github.com/yt-dlp/yt-dlp#update)) or later (specify commit)
- [X] I've checked that all provided URLs are playable in a browser with the same IP and same login details
- [X] I've checked that all URLs and arguments with special characters are [properly quoted or escaped](https://github.com/yt-dlp/yt-dlp/wiki/FAQ#video-url-contains-an-ampersand--and-im-getting-some-strange-output-1-2839-or-v-is-not-recognized-as-an-internal-or-external-command)
- [X] I've searched [known issues](https://github.com/yt-dlp/yt-dlp/issues/3766) and the [bugtracker](https://github.com/yt-dlp/yt-dlp/issues?q=) for similar issues **including closed ones**. DO NOT post duplicates
- [X] I've read the [guidelines for opening an issue](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#opening-an-issue)

### Provide a description that is worded well enough to be understood

Not much to add it just doesn't do the thing

### Provide verbose output that clearly demonstrates the problem

- [X] Run **your** yt-dlp command with **-vU** flag added (`yt-dlp -vU <your command line>`)
- [ ] If using API, add `'verbose': True` to `YoutubeDL` params instead
- [X] Copy the WHOLE output (starting with `[debug] Command-line config`) and insert it below

### Complete Verbose Output

```shell
> yt-dlp -vU
[debug] Command-line config: ['-vU']
[debug] User config "C:\Users\user\AppData\Roaming\yt-dlp\config": ['--ffmpeg-location', 'C:\\Programs\\ffmpeg-for-yt-dlp\\bin\\', '--abort-on-unavailable-fragment', '--no-check-certificate']
[debug] Encodings: locale cp1252, fs utf-8, pref cp1252, out utf-8, error utf-8, screen utf-8
[debug] yt-dlp version stable@2023.07.06 [b532a3481] (win_exe)
[debug] Python 3.8.10 (CPython AMD64 64bit) - Windows-10-10.0.19044-SP0 (OpenSSL 1.1.1k  25 Mar 2021)
[debug] exe versions: ffmpeg N-110472-g4006c71d19-20230504 (setts), ffprobe N-110472-g4006c71d19-20230504
[debug] Optional libraries: Cryptodome-3.18.0, brotli-1.0.9, certifi-2023.05.07, mutagen-1.46.0, sqlite3-2.6.0, websockets-11.0.3
[debug] Proxy map: {}
[debug] Loaded 1855 extractors
[debug] Fetching release info: https://api.github.com/repos/yt-dlp/yt-dlp/releases/latest
ERROR: Unable to obtain version info (HTTP Error 403: rate limit exceeded); Please try again later or visit  https://github.com/yt-dlp/yt-dlp/releases/latest
```

**Comment by bashonly** [comment]
Your IP address is being rate-limited by the github API. 
Have you been running yt-dlp with the `-U` flag repeatedly?
Or is there another process making requests to the github API from your IP?

**Comment by cheater** [comment]
no, not that I know of.

**Comment by cheater** [comment]
is there any way to see a log of those requests? i can't exclude that there's *some* ill-conceived piece of crapware running on my OS, but i wouldn't know what it was.

**Comment by cheater** [comment]
The only thing that comes to mind is to contact GH support to see if they'll send me api request logs, we'll see how that pans out.

**Comment by gamer191** [comment]
Can you access Github (and manually update yt-dlp) through a web browser?
Is reverse-engineering Github's API out of scope for this project?

**Comment by bashonly** [comment]
> Can you access Github (and manually update yt-dlp) through a web browser?

API rate limit would not affect this

> Is reverse-engineering Github's API out of scope for this project?

Not sure exactly what you mean, but doing something like this would likely result in yt-dlp being removed from github

**Comment by bashonly** [comment]
As a workaround, you can use `--update-to` and specify the channel and tag, e.g.
`yt-dlp --update-to stable@2023.09.24`

Sort of inconvenient, but it's the best we can do until we know why you are running into the rate-limit

**Comment by cheater** [comment]
So apparently, the GH API is being used from my IP by something else that throttles it.

However, there's nothing else using it, at least that I can find.

I contacted GH and they just told me to fuck off because apparently if someone's using my computer to run scripts then that person illegally using my computer has privacy concerns or whatever. When using a *public* api.

Any suggestions how to figure out what's happening? I'm not great with network capture tools but my guess is I would need something to that effect.

My only devices are a windows desktop and an android phone. I would welcome suggestions on how to debug this.

But going back to the HTTP 403 rate limit.

There should be a way to get a GH API key and use that with yt-dlp instead, because GH support is clearly uninterested in helping people figure out what's hammering their API.

**Comment by bashonly** [comment]
> I contacted GH 

Github doesn't know the circumstances of your network; you could be on shared wifi, etc, so it makes sense that they cannot share log info with you.

> I'm not great with network capture tools but my guess is I would need something to that effect.

This would only be helpful if it is indeed one of your own devices accessing the GH API without your knowledge

> My only devices are a windows desktop and an android phone. 

There may be only 2 devices on your home network, but you could be behind [NAT](https://en.wikipedia.org/wiki/Network_address_translation), and someone else with the same public IP could be hammering the GH API.

> Any suggestions how to figure out what's happening? 

Firstly, I'd scan your devices for viruses/malware. Maybe disable any browser extensions (besides uBlock Origin).

The GH API rate-limit is hourly, so you could try turning off your desktop and disconnecting your phone from wifi for an hour, and then booting up your computer and running `yt-dlp -U` first thing.

**Comment by cheater** [comment]
Well, if it's something that hammers the api every hour, then it should show up in a network sniffer within an hour, without having to turn my whole computing inside out, i think i should try that first.

**Comment by gamer191** [comment]
> There may be only 2 devices on your home network, but you could be behind [NAT](https://en.wikipedia.org/wiki/Network_address_translation), and someone else with the same public IP could be hammering the GH API.

Good point. Perhaps @cheater should try using -6

**Comment by cheater** [comment]
I'm not behind a NAT - my IPv4 is only owned by me. I confirmed it with the ISP before reaching out to github.

**Comment by cheater** [comment]
so I haven't tried since filing the bug, but I tried now, and yt-dlp updated. So the API wasn't throttled. But it still doesn't solve the issue. I suggest leaving this open and I'll report when this happens again.

**Comment by cheater** [comment]
For one thing, I want to check if using -6 will be useful once this happens again.

**Comment by bashonly** [comment]
Closing this, since:
1. It is no longer reproducible, not even by the OP
2. It is an external issue / not something we can fix

You can continue to report your findings if you want, and this issue can be re-opened if any of your new findings negate points 1 and 2

**Comment by cheater** [comment]
@bashonly sorry, but there's no reason to believe that this bug is permanently unreproducible. it's just not reproducible today. there needs to be a way to run the command without using the anonymous api.

**Comment by bashonly** [comment]
>  there needs to be a way to run the command without using the anonymous api.

There already is, as I mentioned above:

> As a workaround to bypass the GH API, you can use --update-to and specify the channel and tag, e.g.
`yt-dlp --update-to stable@2023.09.24`

`yt-dlp --update-to stable@2023.10.07`

**Comment by cheater** [comment]
@bashonly that's a different command. I'm talking about the -U command. So to restate this clearly: "there needs to be a way to run the -U command without using the anonymous api."

**Comment by Grub4K** [comment]
There doesn't *need* to be. The updater is provided only for convenience and you can always get the update using the api yourself. There you can use whatever token or endpoint you want. Feel free to write a wrapper script for it instead, doing it in yt-dlp itself is ***off-scope***

`--update` is the same as `--update-to latest` and specifying the tag directly skips api requests, which is why it was suggested to you. You can use a single web request to get the current tag, then use `--update-to`:

### `bash`
```bash
yt-dlp --update-to "stable@$(curl https://github.com/yt-dlp/yt-dlp/releases/latest -sI | grep "^location: " | cut -d '/' -f 8)"
```

### `cmd.exe`
```batch
for /f "tokens=7 delims=/" %%v in (
    'curl https://github.com/yt-dlp/yt-dlp/releases/latest -sI ^| findstr /bc:"Location: "'
) do yt-dlp --update-to "stable@%%v"
```
Outside of a batch file you have to use `%v` instead of `%%v`

### PowerShell
```powershell
$Uri = 'https://github.com/yt-dlp/yt-dlp/releases/latest'
$Request = Invoke-WebRequest -Method Get -Uri $Uri -MaximumRedirection 0 -ErrorAction SilentlyContinue
yt-dlp --update-to stable@$($Request.headers.Location.Split('/')[-1])
```

**Comment by cheater** [comment]
so why doesn't yt-dlp do that, instead of using the public github api?

**Comment by Grub4K** [comment]
Bc the public github api is stable while webscraping may break. Take one of the proposed workarounds or don't, either way, like mentioned above
> doing it in yt-dlp itself is ***off-scope***

**Comment by maxugly** [comment]
I just had the same error. Before subbmiting this, I made one last attempt, GH let me in. I am not implying it is a yt-dlp issue, to be clear, just reporting. 

My internet setup is the same os OP, without IPv6. Nothing on my network should be talking to GH at at all besides this tab on my browser and the few attempts I made to update. There is a lot going on on my LAN, admittedly. How likely is this something malicious? I did recently install a lot of XDG/freedesktop.org thumbnailers for tumblerd/thunar but the only one I pointed at a website's API (not GitHub) was for my movie directory. I will go back and do my best to make sense of the source for those. opnsnese does run a cron job to grab blocklists for unbound but I just checked and that ran last 19 hours ago.. **If anything comes up, I will report there and link to it in a new post here** Obviously some malicious or buggy GH-API-spamming process would affect way more than just this project. 

Here is the output, I don't know of a way to show timestamps, at least five to ten minutes went by before last attempt

```
max@minty:~$ yt-dlp -U -v
[debug] Command-line config: ['-U', '-v']
[debug] Encodings: locale UTF-8, fs utf-8, pref UTF-8, out utf-8, error utf-8, screen utf-8
[debug] yt-dlp version stable@2023.11.16 from yt-dlp/yt-dlp [24f827875] (debian*)
[debug] Python 3.10.12 (CPython x86_64 64bit) - Linux-6.2.0-39-generic-x86_64-with-glibc2.35 (OpenSSL 3.0.2 15 Mar 2022, glibc 2.35)
[debug] exe versions: ffmpeg 4.4.2 (setts), ffprobe 4.4.2, rtmpdump 2.4
[debug] Optional libraries: Cryptodome-3.11.0, brotli-1.0.9, certifi-2020.06.20, mutagen-1.45.1, pyxattr-0.7.2, requests-2.22.0, secretstorage-3.3.1, sqlite3-3.37.2, urllib3-1.25.11, websockets-9.1
[debug] Proxy map: {}
[debug] Request Handlers: urllib
[debug] Loaded 1901 extractors
[debug] Fetching release info: https://api.github.com/repos/yt-dlp/yt-dlp/releases/latest
ERROR: Unable to obtain version info (HTTP Error 403: rate limit exceeded); Please try again later or visit  https://github.com/yt-dlp/yt-dlp/releases/latest
max@minty:~$ yt-dlp --update-to "stable@$(curl https://github.com/yt-dlp/yt-dlp/releases/latest -sI | grep "^location: " | cut -d '/' -f 8)"
ERROR: tag/2023.11.16
max@minty:~$ yt-dlp -U -v
[debug] Command-line config: ['-U', '-v']
[debug] Encodings: locale UTF-8, fs utf-8, pref UTF-8, out utf-8, error utf-8, screen utf-8
[debug] yt-dlp version stable@2023.11.16 from yt-dlp/yt-dlp [24f827875] (debian*)
[debug] Python 3.10.12 (CPython x86_64 64bit) - Linux-6.2.0-39-generic-x86_64-with-glibc2.35 (OpenSSL 3.0.2 15 Mar 2022, glibc 2.35)
[debug] exe versions: ffmpeg 4.4.2 (setts), ffprobe 4.4.2, rtmpdump 2.4
[debug] Optional libraries: Cryptodome-3.11.0, brotli-1.0.9, certifi-2020.06.20, mutagen-1.45.1, pyxattr-0.7.2, requests-2.22.0, secretstorage-3.3.1, sqlite3-3.37.2, urllib3-1.25.11, websockets-9.1
[debug] Proxy map: {}
[debug] Request Handlers: urllib
[debug] Loaded 1901 extractors
[debug] Fetching release info: https://api.github.com/repos/yt-dlp/yt-dlp/releases/latest
Latest version: stable@2023.11.16 from yt-dlp/yt-dlp
yt-dlp is up to date (stable@2023.11.16 from yt-dlp/yt-dlp)

```

Your work is inredible and appreciated! Happy holidays yt-dlp!

**Comment by bashonly** [comment]
> [debug] yt-dlp version stable@2023.11.16 from yt-dlp/yt-dlp [24f827875] (debian*)

Just fyi, debian patches yt-dlp to break the self-updater

> How likely is this something malicious?

Can't really say, but I would think that something malicious wouldn't relent enough for you to get a successful response from the github API as you did in your log

**Comment by maxugly** [comment]
> Just fyi, debian patches yt-dlp to break the self-updater

Man, I am not even on Debian. I purposefully wanted something simple, stable and functional for my main PC. Almost anything that can be is offloaded onto a VM on a proxmox node. I am on Mint, and not the Debian version they are cooking up. 

I don't understand python very well, and I think it warned me not to use sudo when I went the pip route. I don't remember exactly. I guess I need to learn venv stuff. I installed via apt. I just looked at my repos. The Official Repositories are Mint and Ubuntu Jammy. I hate adding repos, but I chose that over potentially breaking python. I have done the same for years with Fsearch without issue.  Sorry to bother you with all this.  I had installed "https://ppa.launchpadcontent.net/tomtomtom/yt-dlp/ubuntu/dists/jammy/" and forgot all about it. 

I was under the impression that Ubuntu had strayed away from Debian enough to where it didn't inherit those weird little quirks that come up here and there. Everything has been working great since. Thank you for your time.
